(table): Use ReadonlyArray in 'datasource' property type

#### Bug, feature request, or proposal:
Bug.

#### What is the expected behavior?
`datasource` property should use `ReadonlyArray` instead of `[]`, since it isn't meant to mutate the source array:
```
dataSource: DataSource<T> | Observable<ReadonlyArray<T>> | ReadonlyArray<T>;
```

#### What is the current behavior?
The actual type is `dataSource: DataSource<T> | Observable<T[]> | T[];`.

#### What is the use-case or motivation for changing an existing behavior?
Developers should employ immutability, thus, use `ReadonlyArray`s wherever possible. In this case, we can't just assign such `ReadonlyArray` as a `dataSource` due to a TS error:

> error TS2322: Type 'Observable<ReadonlyArray<T>>' is not assignable to type 'T[] | DataSource<T> | Observable<T[]>'.

This is not a breaking change, since regular non-readonly arrays are still allowed to be assigned as datasources.

#### Which versions of Angular, Material, OS, TypeScript, browsers are affected?
6.4.6

--- a/src/cdk/table/table.ts
+++ b/src/cdk/table/table.ts
@@ -65,6 +65,13 @@ export interface RowOutlet {
 }
 
 /**
+ * Union of the types that can be set as the data source for a `CdkTable`.
+ * @docs-private
+ */
+type CdkTableDataSourceInput<T> = DataSource<T> | Observable<ReadonlyArray<T> | T[]> |
+                                  ReadonlyArray<T> | T[];
+
+/**
  * Provides a handle for the table to grab the view container's ng-container to insert data rows.
  * @docs-private
  */
@@ -310,13 +317,13 @@ export class CdkTable<T> implements AfterContentChecked, CollectionViewer, OnDes
    * subscriptions registered during the connect process).
    */
   @Input()
-  get dataSource(): DataSource<T> | Observable<T[]> | T[] { return this._dataSource; }
-  set dataSource(dataSource: DataSource<T> | Observable<T[]> | T[]) {
+  get dataSource(): CdkTableDataSourceInput<T> { return this._dataSource; }
+  set dataSource(dataSource: CdkTableDataSourceInput<T>) {
     if (this._dataSource !== dataSource) {
       this._switchDataSource(dataSource);
     }
   }
-  private _dataSource: DataSource<T> | Observable<T[]> | T[] | T[];
+  private _dataSource: CdkTableDataSourceInput<T>;
 
   /**
    * Whether to allow multiple rows per data object by evaluating which rows evaluate their 'when'
@@ -759,7 +766,7 @@ export class CdkTable<T> implements AfterContentChecked, CollectionViewer, OnDes
    * render change subscription if one exists. If the data source is null, interpret this by
    * clearing the row outlet. Otherwise start listening for new data.
    */
-  private _switchDataSource(dataSource: DataSource<T> | Observable<T[]> | T[]) {
+  private _switchDataSource(dataSource: CdkTableDataSourceInput<T>) {
     this._data = [];
 
     if (this.dataSource instanceof DataSource) {

